OCPBUGS-44185: Race in configure-ovs.sh affects bonding interface configuration.

djlwilder opened the pull request:
Bonded network configurations with mode=active-backup and fail_over_mac=follow are not functioning due to a race when activating network profiles. activate_nm_connections() attempts to activate all its generated profiles that are not currently in the "active" state. As autoconnect-slaves is set, once br-ex is activated the bond and all its slaves are automatically activated. Their state is set to "activating" until they become active. The "activating" state is not tested for therefor some of the subordinate profiles maybe activated multiple times causing a race in the bonding driver and incorrectly configuring the bond.

Link: https://github.com/openshift/machine-config-operator/issues/4605
Fixes: #4605 

Please provide the following information:
Please see https://github.com/openshift/machine-config-operator/issues/4605
Testing: verify a 2 slave bond with mode=active-backup and fail_over_mac=follow converts to a working OVS config.
The slaves should not have the same MAC.  Other configs should be tested as well.
pull request for inclusion in the changelog: Fix race in configure-ovs.sh that affects bonding interface configuration.

yuqi-zhang commented:
/assign cybertron

@cybertron would you be able to take a look?

djlwilder commented:
/test e2e-azure-ovn-upgrade-out-of-change

cybertron commented:
The change seems fine to me, although I should note that I don't think we do much (if any) testing with the follow mode.

/cc @jcaamano @rbbratta

djlwilder commented:
Hi All,  thanks for reviewing my PR.  Please let me know if you need further clarification on the issue or the fix.  The debug traces I attached to issue 4605 should help explain the interaction between the NetworkManager and the bonding driver when  configure_ovs.sh is executed.

prb112 commented:
/retest-required

prb112 commented:
/retest-required

prb112 commented:
Hey @jcaamano @rbbratta if you have feedback, that'd be great. Many thanks, Paul

jcaamano commented:
@djlwilder 

Can you help me understand? If the profiles that we want active are active, and the profiles we want inactive are inactive, how is this a race in configure-ovs? If the observable state from configure-ovs matches what it wants, isn't the race in NetworkManager or some other place?

There is two things that are being taken care of here that would be at risk with this change:

- activation order: unless we follow a specific order, some unwanted profiles might be activated as well. We can only control the order via explicitly activating.
- ensuring things are active before configure-ovs quits

Unwanted profiles generally refers to profiles generated from kargs or leftovers from a previous boot if configure-ovs is told to use `/etc/NetworkManager/system-connections` instead of `/run/NetworkManager/system-connections`. The latter is not default and improbable because it is also not documented, but the former might be a more common issue. So have you tested two reboots with the bond originally defined though kargs?

There is also some other considerations:
- what if the profile gets to `activating` state one instant after we check for it? 
- if ovs-if-phys0 has auto-connect slaves set, meaning the slaves can be re-activated when ovs-if-phys0 is activated, wouldn't the same thing happen?

Overall, this looks to me as not a race in configure-ovs so not something that we can absolutely fix in configure-ovs. If we are going to mitigate here, I would have two asks:
- Learning more about what is the actual race condition. You mentioned in the issue `These interfaces are not activated a second time leaving the state of the bond in an unpredictable state` talking about the slave profiles. What does this exactly mean? It doesn't make sense to me because that is what you are actually preventing in this PR: for the profiles to be activated a second time. We are actually making sure the profiles we need active are active in the order we want, to avoid...well you guessed it, other race conditions.
- crosschecking at the end that all profiles are activated

djlwilder commented:
@jcaamano

> 
> Can you help me understand? If the profiles that we want active are active, and the profiles we want inactive are inactive, how is this a race in configure-ovs? If the observable state from configure-ovs matches what it wants, isn't the race in NetworkManager or some other place?

Let me describe what I found in more detail so we can get on the same page.

In my setup activate_nm_connections is called with the following list of profiles:
 br-ex ovs-if-phys0 enP32807p1s0-slave-ovs-clone enP49154p1s0-slave-ovs-clone ovs-if-br-ex

I have added some debug to show the state that NetworkManager reports for each profile as we process the list.

Here is a log showing the issue:
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=br-ex State=None  (1)
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=ovs-if-phys0 State=None (1)
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=enP32807p1s0-slave-ovs-clone State=None (1)
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=enP49154p1s0-slave-ovs-clone State=None (1)
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=ovs-if-br-ex State=None (1)
DEBUG-ShowNMStates: Attempt 1 to bring up connection br-ex
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=ovs-if-phys0 State=activating (2)
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=enP32807p1s0-slave-ovs-clone State=activating (2)
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=enP49154p1s0-slave-ovs-clone State=activating (2)
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: Attempt 1 to bring up connection ovs-if-phys0  (3)
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-phys0 successfully conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-phys0 successfully conn=ovs-if-phys0 State=activating
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-phys0 successfully conn=enP32807p1s0-slave-ovs-clone State=activating
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-phys0 successfully conn=enP49154p1s0-slave-ovs-clone State=activating
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-phys0 successfully conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: Attempt 1 to bring up connection enP32807p1s0-slave-ovs-clone (3)
DEBUG-ShowNMStates: DebugLoc=Brought up connection enP32807p1s0-slave-ovs-clone successfully conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection enP32807p1s0-slave-ovs-clone successfully conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection enP32807p1s0-slave-ovs-clone successfully conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection enP32807p1s0-slave-ovs-clone successfully conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection enP32807p1s0-slave-ovs-clone successfully conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=br-ex State=activated (4)
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: Attempt 1 to bring up connection ovs-if-br-ex
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=ovs-if-br-ex State=activated

(1) At the start of activate_nm_connections() profiles have no state (shown as None)

(2) After br-ex is activated (nmcli conn up), ovs-if-phys0, enP32807p1s0-slave-ovs-clone and enP49154p1s0-slave-ovs-clone have a new state of "activating".  Presumably NM is working to move these profiles to "active".

(3) As the state of "activating" is not checked ovs-if-phys0 and enP32807p1s0-slave-ovs-clone are activated as well (nmcli conn up).

(4) However, by the time we check the state of enP49154p1s0-slave-ovs-clone is already activated so no action is taken.

The timing here is critical, in this next test I added a "sleep 1" after br-ex is reported as active. 
By giving NM more time NM was able to completed activating: ovs-if-phys0, enP32807p1s0-slave-ovs-clone, and enP49154p1s0-slave-ovs-clone so no additional activation's were needed.  This is the race I spoke of.

DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=br-ex State=None
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=ovs-if-phys0 State=None
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=enP32807p1s0-slave-ovs-clone State=None
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=enP49154p1s0-slave-ovs-clone State=None
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: Attempt 1 to bring up connection br-ex
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=ovs-if-phys0 State=activating
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=enP32807p1s0-slave-ovs-clone State=activating
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=enP49154p1s0-slave-ovs-clone State=activating
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: Sleeping for 1 sec.  **********
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated. conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated. conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated. conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated. conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated. conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated. conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated. conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated. conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated. conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated. conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated. conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: Attempt 1 to bring up connection ovs-if-br-ex
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=ovs-if-br-ex State=activated


Finally, here is debug output with my change to configure-ovs.sh (checking for state=activating).

DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=br-ex State=None
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=ovs-if-phys0 State=None
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=enP32807p1s0-slave-ovs-clone State=None
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=enP49154p1s0-slave-ovs-clone State=None
DEBUG-ShowNMStates: DebugLoc=top-of activate_nm_connections conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: Attempt 1 to bring up connection br-ex
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=ovs-if-phys0 State=activating
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=enP32807p1s0-slave-ovs-clone State=activating
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=enP49154p1s0-slave-ovs-clone State=activating
DEBUG-ShowNMStates: DebugLoc=Brought up connection br-ex successfully conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated or activating. conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated or activating. conn=ovs-if-phys0 State=activating
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated or activating. conn=enP32807p1s0-slave-ovs-clone State=activating
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated or activating. conn=enP49154p1s0-slave-ovs-clone State=activating
DEBUG-ShowNMStates: DebugLoc=Connection ovs-if-phys0 already activated or activating. conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated or activating. conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated or activating. conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated or activating. conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated or activating. conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP32807p1s0-slave-ovs-clone already activated or activating. conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated or activating. conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated or activating. conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated or activating. conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated or activating. conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Connection enP49154p1s0-slave-ovs-clone already activated or activating. conn=ovs-if-br-ex State=None
DEBUG-ShowNMStates: Attempt 1 to bring up connection ovs-if-br-ex
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=br-ex State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=ovs-if-phys0 State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=enP32807p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=enP49154p1s0-slave-ovs-clone State=activated
DEBUG-ShowNMStates: DebugLoc=Brought up connection ovs-if-br-ex successfully conn=ovs-if-br-ex State=activated

So the problem is not if the profiles are active or not its how you got there. I suspect that by setting autoconnect-slaves=1 we are allowing NM
to activate some profiles for us thus removing our ability to manage then (I need to verify this).  

The next question is why this is a problem.  The debug traces I attached to issue 4605 should help explain hoe this affects the bonding driver.  

Thank you for your help.

djlwilder commented:
> Can you help me understand? If the profiles that we want active are active, and the profiles we want inactive are inactive, how is this a race in configure-ovs? If the observable state from configure-ovs matches what it wants, isn't the race in NetworkManager or some other place?
> 
> There is two things that are being taken care of here that would be at risk with this change:
> 
>     * activation order: unless we follow a specific order, some unwanted profiles might be activated as well. We can only control the order via explicitly activating.

Anytime the NM has set a "activating" state we may have lost control over activation order already. 
   
> 
>     * ensuring things are active before configure-ovs quits
I agree

> 
> 
> Unwanted profiles generally refers to profiles generated from kargs or leftovers from a previous boot if configure-ovs is told to use `/etc/NetworkManager/system-connections` instead of `/run/NetworkManager/system-connections`. The latter is not default and improbable because it is also not documented, but the former might be a more common issue. So have you tested two reboots with the bond originally defined though kargs?

I have not tested with kargs,  will do.
> 
> There is also some other considerations:
> 
>     * what if the profile gets to `activating` state one instant after we check for it?

I agree, we have no control how long a NM will keep a profile in `activating` state so that is going to be a problem.  Finding a solution that is not dependent on a `activating` state would be ideal.
> 
>     * if ovs-if-phys0 has auto-connect slaves set, meaning the slaves can be re-activated when ovs-if-phys0 is activated, wouldn't the same thing happen?

Are you saying if "nmcli con up ovs-if-phys0" is run some point after configure-ovs is run?  I definitely need to test this.  Although I only see it as an issue if we follow up by activating the slaves one at a time as the script will do.  (see below for more on this).

> 
> 
> Overall, this looks to me as not a race in configure-ovs so not something that we can absolutely fix in configure-ovs. If we are going to mitigate here, I would have two asks:
> 
>     * Learning more about what is the actual race condition. You mentioned in the issue `These interfaces are not activated a second time leaving the state of the bond in an unpredictable state` talking about the slave profiles. What does this exactly mean? It doesn't make sense to me because that is what you are actually preventing in this PR: for the profiles to be activated a second time. We are actually making sure the profiles we need active are active in the order we want, to avoid...well you guessed it, other race conditions.

 I hope this better explain the problem. In my example the problem was triggered because one  *-slave-ovs-clone was explicitly activated and the other was not this is due to the unpredictable delay between activating->activated. If the timing was such that neither or both were explicitly activated the outcome would be different.  (see also note above)

 Here is another approach that might address your concerns:

-  I am not checking for state="activating" to prevent the issue with checking reading the state to early.
-  I follow the order of profiles,  for profiles that can have a state of activating just wait untill the state="activated".

```
    # Activate all interfaces that are not yet active
    # NetworkManager will activate slave interfaces for us once the slave's
    # parent has been activated. So don't activate slaves, just wait for them to become
    # active. If slave done become active after a suficent time give up and move on.
    # TBD: this works with slave_type=bond,  what about slave_type=team?

    for i in {1..10}; do
      echo "Attempt $i to bring up connection $conn"
  
      if $is_slave; then
        active_state=$(nmcli -g GENERAL.STATE conn show "$conn")
        if [ "$active_state" == "activated" ]; then
           s=0
           break
        fi
      else
           nmcli conn up "$conn" && s=0 && break || s=$?
      fi
      sleep 5
    done
    if [ $s -eq 0 ]; then
      echo "Brought up connection $conn successfully"
      if $is_slave; then
        master_interfaces["$master_interface"]=true
      fi
    elif ! $is_slave; then
      echo "ERROR: Cannot bring up connection $conn after $i attempts"
      return $s
    fi
    mod_nm_conn "$conn" connection.autoconnect yes
```
 
> 
>     * crosschecking at the end that all profiles are activated
Ok,  what should I do if some profiles don't get activated?

djlwilder commented:
> I guess this is my question as well and thus my statement that the race must be somewhere else as activating a profile multiple times should not be a problem. However this is not a question we can answer, rather the RHEL team should. Anyway, I recognize the need to avoid the problem here as much as we can.

I am all for finding root cause.  In this case I think that root cause is a limitation in the bonding drive coupled with how network manager interacts with it.   I can open a bugzilla against RHEL for this.  If you agree I would like to continue the path of avoiding the issue as long as we don't introduce more problems.  I like your latest change suggestion,  this should prevent the problem I am seeing.  Let me experiment with this and I will post code and results soon.

djlwilder commented:
**I updated my fork with the changes as we discussed  and ran unit tests with bonding configured and fail_over_mac=follow. Platform: PowerVM**
```
# oc version
Client Version: 4.16.20
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: 4.16.20
Kubernetes Version: v1.29.9+5865c5b
```

Unit tests:
1) Configure bond with kargs:
kargs:append=bond=bond0:enP32771p1s0,enP49154p1s0:mode=active-backup,fail_over_mac=follow ip=192.168.42.6::192.168.42.1:255.255.255.0::bond0:none:192.168.42.1
Booted node, verified that the bond configured correctly and node became ready.

2) Rebooted the node twice using kargs:
No issues seen with either boot. Verified that the bond configured correctly and node became ready.

3) Configure bond by injecting files to /etc/NetworkManager/system-connections 
- Manually add files 
$ ls /etc/NetworkManager/system-connections/
 bond0.nmconnection  enP32771p1s0.nmconnection  enP49154p1s0.nmconnection
- Remove files from /run/NetworkManager/system-connections/
- reboot node
No issues seen. Verified that the bond configured correctly and node became ready.
- Verified generated configuration:
$ ls  /run/NetworkManager/system-connections/
bond0.nmconnection                         enP49154p1s0-slave-ovs-clone.nmconnection  ovs-if-phys0.nmconnection
br-ex.nmconnection                         enP49154p1s0.nmconnection                  ovs-port-br-ex.nmconnection
enP32771p1s0-slave-ovs-clone.nmconnection  lo.nmconnection                            ovs-port-phys0.nmconnection
enP32771p1s0.nmconnection                  ovs-if-br-ex.nmconnection

4) Missing slave test.
- Removed one slave from from node's configuration (hyper-visor configuration).
- Booted node.
No issues seen. Verified that the bond configured correctly with a single slave and node became ready.

5) Bond fail-over test.
Booted node, verified bond is configured and working.
```
$ cat /sys/class/net/bond0/bonding/slaves
enP32771p1s0 enP49154p1s0

$ cat /sys/class/net/bond0/bonding/active_slave 
enP32771p1s0

$ echo enP49154p1s0 >/sys/class/net/bond0/bonding/active_slave

$ cat /sys/class/net/bond0/bonding/active_slave
enP49154p1s0
```
Verified bond0 is functional and slave MACS have swapped.

djlwilder commented:
/retest-required

djlwilder commented:
/test e2e-azure-ovn-upgrade-out-of-change

prb112 commented:
/test e2e-gcp-op-ocl

djlwilder commented:
/test e2e-gcp-op-ocl

djlwilder commented:
> > I guess this is my question as well and thus my statement that the race must be somewhere else as activating a profile multiple times should not be a problem. However this is not a question we can answer, rather the RHEL team should. Anyway, I recognize the need to avoid the problem here as much as we can. 

 @jcaamano
I reproduced the problem outside of OpenShift on RHEL9.6 and with upstream net-next kernel.  I filled this bug against RHEL 9.6.
RHEL-73616 - RHEL9.6: Using nmcli to activate (up) or deactivate (down) the active slaves breaks the bond.

Have you had a chance to look at my updated commit?

Thank you.

jcaamano commented:
/lgtm

sherine-k commented:
/check-required-labels

rbbratta commented:
there have been NM changes to autoactivation for a different race from [NetworkManager-1.46.0-26.el9_4](https://brewweb.engineering.redhat.com/brew/buildinfo?buildID=3510355)

- Connection autoactivation gets blocked after failing (RHEL-78747)

https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/2136

djlwilder commented:
> there have been NM changes to autoactivation for a different race in [NetworkManager-1.46.0-26.el9_4](https://brewweb.engineering.redhat.com/brew/buildinfo?buildID=3510355)
> 
>     * Connection autoactivation gets blocked after failing ([RHEL-78747](https://issues.redhat.com//browse/RHEL-78747))
> 
> 
> https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/2136

I don't believe this will affect  the change in this PR. The patch talks about resetting a retry count on a internal activation to avoid blocking the NICs after failing activation due to ovs-bond not ready yet.   The problem in this PR is that  NM was activating a bond-slave twice,  once by the internal activation once explicitly by running "nmcli con up".   The second activation was not needed and causing the problem.  My change waits for the internal activation to complete then skips the "nmcli con up" .  The NM change may affect the timing of the activation but the internal activation should still happen.    So this PR is still valid the change is still needed.

djlwilder commented:
> > there have been NM changes to autoactivation for a different race in [NetworkManager-1.46.0-26.el9_4](https://brewweb.engineering.redhat.com/brew/buildinfo?buildID=3510355)
> > ```
> > * Connection autoactivation gets blocked after failing ([RHEL-78747](https://issues.redhat.com//browse/RHEL-78747))
> > ```
> > https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/2136
> 
> I don't believe this will affect the change in this PR. The patch talks about resetting a retry count on a internal activation to avoid blocking the NICs after failing activation due to ovs-bond not ready yet. The problem in this PR is that NM was activating a bond-slave twice, once by the internal activation once explicitly by running "nmcli con up". The second activation was not needed and causing the problem. My change waits for the internal activation to complete then skips the "nmcli con up" . The NM change may affect the timing of the activation but the internal activation should still happen. So this PR is still valid the change is still needed.

I updated the NetworkManager rpms to 1.46.0-26.el9_4. (the version with the above change) and confirmed:
1) The problem described in this PR was not solved by the update.
2) The change (fix) described in this PR still worked to prevent the problem after NM was updated.

prb112 commented:
/retest-required

djlwilder commented:
/retest-required

yuqi-zhang reviewed (approved):
Leaving an approval from the MCO side.

We did a configuration change that inadvertantly turned on some old tests, feel free to ignore for now

prb112 commented:
/retest-required

prb112 commented:
/retest-required

yuqi-zhang commented:
there are some general CI issues and MCO CI issues at this time, will see if that resolves

prb112 commented:
/retest-required

sherine-k commented:
/retest-required

prb112 commented:
FYI, I see the job `ci/prow/4.12-upgrade-from-stable-4.11-images` failing for other PRs I'm following.

prb112 commented:
/retest-required

lzarnerg commented:
We have tried this fix: worked fine with failover from live Active=>Backup IF, but fail-back from running Backup=>Active IF produces same error (which appears to be that configure-ovs.sh keeps running with result that same MAC appears on both sides of local switches => no Ethernet delivery). Short-term fix, disable L2 on what was the live Backup IF and Ethernet delivery resumes.
